src/Razor/src/Microsoft.VisualStudio.RazorExtension/SyntaxVisualizer/SourceMappingTagger.cs

```diff
@@ -0,0 +1,128 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.VisualStudio.Editor.Razor.Documents;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Tagging;
+
+namespace Microsoft.VisualStudio.RazorExtension.SyntaxVisualizer
+{
+    internal sealed class SourceMappingTagger : ITagger<SourceMappingTag>
+    {
+        internal static SourceMappingTagger GetOrCreateTagger(ITextBuffer buffer, Func<SourceMappingTagger> creator)
+        {
+            return buffer.Properties.GetOrCreateSingletonProperty(creator);
+        }
+
+        private readonly ITextBuffer _buffer;
+        private readonly Lazy<RazorCodeDocumentProvidingSnapshotChangeTrigger> _sourceMappingProjectChangeTrigger;
+        private readonly ITextDocumentFactoryService _textDocumentFactoryService;
+
+        public event EventHandler<SnapshotSpanEventArgs>? TagsChanged;
+
+        public static bool Enabled { get; set; }
+
+        internal SourceMappingTagger(ITextBuffer buffer, Lazy<RazorCodeDocumentProvidingSnapshotChangeTrigger> sourceMappingProjectChangeTrigger, ITextDocumentFactoryService textDocumentFactoryService)
+        {
+            _buffer = buffer;
+            _sourceMappingProjectChangeTrigger = sourceMappingProjectChangeTrigger;
+            _textDocumentFactoryService = textDocumentFactoryService;
+            _buffer.Changed += (sender, args) => HandleBufferChanged(args);
+        }
+
+        public IEnumerable<ITagSpan<SourceMappingTag>> GetTags(NormalizedSnapshotSpanCollection spans)
+        {
+            if (!Enabled || spans.Count == 0)
+            {
+                return Enumerable.Empty<ITagSpan<SourceMappingTag>>();
+            }
+
+            var snapshot = spans[0].Snapshot;
+
+            if (!_textDocumentFactoryService.TryGetTextDocument(_buffer, out var textDocument))
+            {
+                return Enumerable.Empty<ITagSpan<SourceMappingTag>>();
+            }
+
+            var codeDocument = ThreadHelper.JoinableTaskFactory.Run(() => _sourceMappingProjectChangeTrigger.Value.GetRazorCodeDocumentAsync(textDocument.FilePath, CancellationToken.None));
+
+            if (codeDocument is null)
+            {
+                return Enumerable.Empty<ITagSpan<SourceMappingTag>>();
+            }
+
+            return GetTagsWorker();
+
+            IEnumerable<ITagSpan<SourceMappingTag>> GetTagsWorker()
```

Comment:
Do we want this to be a capturing function or could we make it static and pass `codeDocument` and `csharpDocument`?

Reply:
I'm not fussed about it at all, but I can change it if you prefer